File: adapters/backend/php/src/RampartClaims.php
<?php

declare(strict_types=1);

namespace Rampart\Laravel;

/**
 * Verified JWT claims from a Rampart token.
 *
 * Mirrors the claims structure used across all Rampart adapter SDKs.
 */
final class RampartClaims
{
    /**
     * @param string      $sub               Subject (user ID)
     * @param string      $iss               Issuer URL
     * @param int         $iat               Issued-at timestamp
     * @param int         $exp               Expiration timestamp
     * @param string|null $orgId             Organization / tenant ID
     * @param string|null $preferredUsername  Username
     * @param string|null $email             Email address
     * @param bool|null   $emailVerified     Whether email is verified
     * @param string|null $givenName         First name
     * @param string|null $familyName        Last name
     * @param string[]    $roles             Assigned roles
     */
    public function __construct(
        public readonly string $sub,
        public readonly string $iss,
        public readonly int $iat,
        public readonly int $exp,
        public readonly ?string $orgId = null,
        public readonly ?string $preferredUsername = null,
        public readonly ?string $email = null,
        public readonly ?bool $emailVerified = null,
        public readonly ?string $givenName = null,
        public readonly ?string $familyName = null,
        public readonly array $roles = [],
    ) {}

    /**
     * Build a RampartClaims instance from a decoded JWT payload.
     *
     * @param object|array<string, mixed> $payload Decoded JWT payload
     * @return self
     */
    public static function fromToken(object|array $payload): self
    {
        $data = (array) $payload;

        $roles = $data['roles'] ?? [];
        if (is_string($roles)) {
            $roles = [$roles];
        }

        return new self(
            sub: (string) ($data['sub'] ?? ''),
            iss: (string) ($data['iss'] ?? ''),
            iat: (int) ($data['iat'] ?? 0),
            exp: (int) ($data['exp'] ?? 0),
            orgId: isset($data['org_id']) ? (string) $data['org_id'] : null,
            preferredUsername: isset($data['preferred_username']) ? (string) $data['preferred_username'] : null,
            email: isset($data['email']) ? (string) $data['email'] : null,
            emailVerified: isset($data['email_verified']) ? (bool) $data['email_verified'] : null,
            givenName: isset($data['given_name']) ? (string) $data['given_name'] : null,
            familyName: isset($data['family_name']) ? (string) $data['family_name'] : null,
            roles: array_values(array_map('strval', (array) $roles)),
        );
    }

    /**
     * Check whether this token has ALL of the given roles.
     *
     * @param string ...$requiredRoles
     * @return bool
     */
    public function hasRoles(string ...$requiredRoles): bool
    {
        foreach ($requiredRoles as $role) {
            if (!in_array($role, $this->roles, true)) {
                return false;
            }
        }
        return true;
    }

    /**
     * Return the claims that are missing from the token.
     *
     * @param string ...$requiredRoles
     * @return string[]
     */
    public function missingRoles(string ...$requiredRoles): array
    {
        $missing = [];
        foreach ($requiredRoles as $role) {
            if (!in_array($role, $this->roles, true)) {
                $missing[] = $role;
            }
        }
        sort($missing);
        return $missing;
    }

    /**
     * Convert claims to an associative array.
     *
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        return [
            'sub' => $this->sub,
            'iss' => $this->iss,
            'iat' => $this->iat,
            'exp' => $this->exp,
            'org_id' => $this->orgId,
            'preferred_username' => $this->preferredUsername,
            'email' => $this->email,
            'email_verified' => $this->emailVerified,
            'given_name' => $this->givenName,
            'family_name' => $this->familyName,
            'roles' => $this->roles,
        ];
    }
}

File: adapters/backend/dotnet/RampartClaims.cs
using System.Security.Claims;

namespace Rampart.AspNetCore;

/// <summary>
/// Typed representation of Rampart JWT claims.
/// Use <see cref="FromPrincipal"/> to extract claims from a validated <see cref="ClaimsPrincipal"/>.
/// </summary>
public sealed class RampartClaims
{
    /// <summary>Subject identifier (user ID).</summary>
    public string Sub { get; init; } = string.Empty;

    /// <summary>Organization ID.</summary>
    public string? OrgId { get; init; }

    /// <summary>Username.</summary>
    public string? PreferredUsername { get; init; }

    /// <summary>Email address.</summary>
    public string? Email { get; init; }

    /// <summary>Whether the email address has been verified.</summary>
    public bool EmailVerified { get; init; }

    /// <summary>First name (if set).</summary>
    public string? GivenName { get; init; }

    /// <summary>Last name (if set).</summary>
    public string? FamilyName { get; init; }

    /// <summary>Assigned roles.</summary>
    public IReadOnlyList<string> Roles { get; init; } = Array.Empty<string>();

    /// <summary>
    /// Extracts typed Rampart claims from a validated <see cref="ClaimsPrincipal"/>.
    /// </summary>
    /// <param name="principal">The claims principal from the authenticated user.</param>
    /// <returns>A <see cref="RampartClaims"/> instance with extracted claim values.</returns>
    /// <exception cref="ArgumentNullException">Thrown when <paramref name="principal"/> is null.</exception>
    public static RampartClaims FromPrincipal(ClaimsPrincipal principal)
    {
        ArgumentNullException.ThrowIfNull(principal, nameof(principal));

        var identity = principal.Identity as ClaimsIdentity;
        if (identity == null)
        {
            return new RampartClaims();
        }

        var roles = principal.FindAll(ClaimTypes.Role)
            .Select(c => c.Value)
            .Concat(principal.FindAll("roles").Select(c => c.Value))
            .Distinct()
            .ToList();

        var emailVerifiedClaim = identity.FindFirst("email_verified");
        var emailVerified = emailVerifiedClaim != null &&
            bool.TryParse(emailVerifiedClaim.Value, out var ev) && ev;

        return new RampartClaims
        {
            Sub = identity.FindFirst(ClaimTypes.NameIdentifier)?.Value
                  ?? identity.FindFirst("sub")?.Value
                  ?? string.Empty,
            OrgId = identity.FindFirst("org_id")?.Value,
            PreferredUsername = identity.FindFirst("preferred_username")?.Value,
            Email = identity.FindFirst(ClaimTypes.Email)?.Value
                    ?? identity.FindFirst("email")?.Value,
            EmailVerified = emailVerified,
            GivenName = identity.FindFirst(ClaimTypes.GivenName)?.Value
                        ?? identity.FindFirst("given_name")?.Value,
            FamilyName = identity.FindFirst(ClaimTypes.Surname)?.Value
                         ?? identity.FindFirst("family_name")?.Value,
            Roles = roles.AsReadOnly(),
        };
    }

    /// <inheritdoc />
    public override string ToString()
    {
        return $"RampartClaims{{Sub='{Sub}', Email='{Email}', Roles=[{string.Join(", ", Roles)}]}}";
    }
}
